python/ray/dag/tests/experimental/test_execution_schedule.py

```diff
@@ -0,0 +1,271 @@
+# coding: utf-8
+import os
+import sys
+
+import pytest
+
+import ray
+import ray.cluster_utils
+from ray.experimental.channel.torch_tensor_type import TorchTensorType
+from ray.experimental.channel.conftest import start_nccl_mock
+from ray.tests.conftest import *  # noqa
+from ray.dag import InputNode, MultiOutputNode
+from ray.dag.compiled_dag_node import DAGNodeOperationType
+
+if sys.platform != "linux" and sys.platform != "darwin":
+    pytest.skip("Skipping, requires Linux or Mac.", allow_module_level=True)
+
+
+@ray.remote(num_cpus=0, num_gpus=1)
+class MockedWorker:
+    def __init__(self):
+        pass
+
+    def start_mock(self):
+        """
+        Patch methods that require CUDA.
+        """
+        start_nccl_mock()
+
+    def fwd(self, value):
+        return value
+
+    def bwd(self, value):
+        return value
+
+    def read_input(self, input):
+        return input
+
+    def no_op(self, value):
+        return value
+
+    def no_op_two(self, value1, value2):
+        return value1, value2
+
+
+@pytest.mark.parametrize("ray_start_regular", [{"num_gpus": 2}], indirect=True)
+def test_simulate_pp_2workers_1f1b(ray_start_regular, monkeypatch):
+    """
+    This test simulates a simple 1F1B pipeline parallelism for training with
+    2 workers and 2 batches.
+
+    w1: fwd_b1  fwd_b2          bwd_b1          bwd_b2
+    w2:         fwd_b1  bwd_b1  fwd_b2  bwd_b2
+
+    The communication between workers is done using NCCL. The communication
+    within the worker actor is done using IntraProcessChannel.
+    """
+    monkeypatch.setattr(ray.dag.constants, "RAY_ADAG_ENABLE_DETECT_DEADLOCK", False)
```

Comment:
I still need time to think about the deadlock detection in this PR. The existing deadlock detection generates some false alarms after this PR.